share/wake/lib/core/json.wake

```diff
@@ -538,3 +538,162 @@ export def lastValueInList (values: List JValue): Result JValue Error =
+# about only giving jNumber as an abstraction over jDouble...IDK yet.
+export def jDouble: JValue => List Double =
+    def view j =
+        require (JDouble jd) = j else Nil
+        jd, Nil
+
+    view
+
+export def jeDouble: (Double => Double) => JValue => JValue =
+    def edit f j =
+        require (JDouble jd) = j else j
+        JDouble (f jd)
+
+    edit
+
+export def jNumber: JValue => List Double =
+    def view = match _
+        JDouble x  = x, Nil
+        JInteger x = fromOption $ dint x
+        _          = Nil
+
+    view
+
+# This is odd because if you edit a JInteger with say +1 you still get a JDouble
+export def jeNumber: (Double => Double) => JValue => JValue =
+    def edit f s = match s
+        JDouble x  = JDouble (f x)
+        JInteger x =
+            require Some dx = dint x else s
+            JDouble (f dx)
+        _          = s
+
+    edit
+
+export def jNull: JValue => List Unit =
+    def view j =
+        require JNull = j else Nil
+        Unit, Nil
+
+    view
+
+export def jeNull: (Unit => Unit) => JValue => JValue =
+    def edit f j =
+        require JNull = j else j
+        def _ = f Unit # the edit function might be effectful
+        JNull
+
+    edit
+
+export def jObject: JValue => List (List (Pair String JValue)) =
+    def view j =
+        require (JObject jo) = j else Nil
+        jo, Nil
+
+    view
+
+export def jeObject: (List (Pair String JValue) => List (Pair String JValue)) => JValue => JValue =
+    def edit f j =
+        require (JObject jo) = j else j
+        JObject (f jo)
+
+    edit
+
+export def jArray: JValue => List (List JValue) =
+    def view j =
+        require (JArray ja) = j else Nil
+        ja, Nil
+
+    view
+
+export def jeArray: (List JValue => List JValue) => JValue => JValue =
+    def edit f j =
+        require (JArray jl) = j else j
+        JArray (f jl)
+
+    edit
+
+export def jSingelton: JValue => List JValue =
+    def view j =
+        require (JArray (j2, Nil)) = j else Nil
+        j2, Nil
+
+    view
+
+export def jeSingelton: (JValue => JValue) => JValue => JValue =
+    def edit f j =
+        require (JArray (j2, Nil)) = j else j
+        JArray (f j2, Nil)
+
+    edit
+
+export def jAll: JValue => List JValue = _ | jArray |> (_)
+export def jeAll: (JValue => JValue) => JValue => JValue = jeArray $ map $ _
+
+export def jAt (i: Integer): JValue => List JValue = _ | jArray |> atT i
+export def jeAt (i: Integer): (JValue => JValue) => JValue => JValue = jeArray $ atE i $ _
+
+export def jField (field: String): JValue => List JValue =
+    _ | jObject |> lookup (_==~_) field
+
+export def jeField (field: String): (JValue => JValue) => JValue => JValue =
+    jeObject $ lookupE (_==~_) field $ _
+
+# NOTE: Feels like we do need a way to map functions, `... | map f` doesn't work
+export def jFieldMatch (field: RegExp): JValue => List JValue =
+  _ | jObject |> filter (matches field _.getPairFirst) | map getPairSecond
```

Comment:
I realize this is more typing, but it constructs 1 list instead of 3. wake cannot inline through loops.
jField is going to be one of the most heavily used operations.

```
export def jFieldMatch (field: RegExp): JValue => List JValue =
  def pick (Pair key value) =
    if matches field key then Some value else None
  match _
    JObject l = mapPartial pick
    x = Nil
```

Even with this change, `|> jField "xyz"` still does double work, because it has to concat the results of the inner loop. It makes me almost wish that these edit functions took a tail as an input...

Reply:
A few things

1) jField != jFieldMatch. My hope is that jFieldMatch is less used now that we have jField
2) All for optimizing the most used things but I want to limit it to the most used things
3) I never really considered this new mechanism to be for performance minded things. If it is I'd like to maybe consider adding loop fusion now so that by and large we can expect these queries to be efficient. It sounds like a lot of work for not much gain IMO though; not that it wouldn't improve performance but it doesn't seem like it would improve performance *enough* to be worth the time it would take to implement. If you think its easy though maybe we just add it?

Reply:
1. I agree that `jField` is the important case to perform well. However, that case can ALSO return multiple matches, my comment was meant to apply to both methods.
2. Their implementations should be presumably analogous.
3. Loop fusion could be interesting, but IMO other optimizations (ie: SSA2) should probably happen first. Writing code to use less loops upfront in situations where it also decreases the cognitive load of reading the code, though, ... that's just win-win and doesn't require adding new optimization passes. If you can get loop fusion to combine the `|>` and the `jField` automatically ... that would be pretty awesome, though.

How about?:
```
export def jFieldMatch (field: RegExp): JValue => List JValue = match _
  JObject pairs = lookup (matches field) pairs
  _ = Nil
```
... still one loop and eliminates most of the code.

Reply:
I updated both jField and jFieldMatch to this style. The new lookup semantics (e.g. not stopping at the first one plus the better type signature) makes this look a lot cleaner this way IMO.

Part of me does really like composing things out of the building blocks we made though.